primitives/header-chain/src/lib.rs

```diff
@@ -110,3 +112,107 @@ impl<H, P> AncestryChecker<H, P> for () {
 		true
 	}
 }
+
+/// A simple ancestry checker which verifies ancestry by walking every header between `child` and
+/// `ancestor`.
+pub struct LinearAncestryChecker;
+
+impl<H: HeaderT> AncestryChecker<H, Vec<H>> for LinearAncestryChecker {
+	fn are_ancestors(ancestor: &H, child: &H, proof: &Vec<H>) -> bool {
+		// You can't be your own parent
+		if ancestor == child {
```

Comment:
Given all other checks this can be easily optimised into:
```suggestion
		if proof.len() < 2 {
		    return false;
		}
```
which should be cheaper than exact header comparison